### PR TITLE
Callgraph

### PR DESCRIPTION
**Pull request by Geoffrey1014**
add a copy constuctor for CallGraph class
bullding callgraphs in CallGraphBuilder
building threadCallGraphs in ThreadCallGraphBuilder
merge ThreadCallGraphBuilder into CallgraphBuilder
remove callgraph field in CallgraphBuilder

### CODE REVIEW

File: svf/include/Util/CallGraphBuilder.h [hunk 1]
```diff
@@ -38,34 +38,27 @@ namespace SVF
...
 
 class ICFG;
+class SVFModule;
 
 class CallGraphBuilder
 {
-
 protected:
-    CallGraph* callgraph;
+    const SVFModule::FunctionSetType* svfFunctionSet;
```
Comment (jumormt) [inline svf/include/Util/CallGraphBuilder.h:46]: typedef SVFModule::FunctionSetType FunctionSetType;

File: svf/lib/Util/CallGraphBuilder.cpp [hunk 2]
```diff
@@ -56,28 +54,25 @@ CallGraph* CallGraphBuilder::buildCallGraph(SVFModule* svfModule)
...
 }
 
-CallGraph* ThreadCallGraphBuilder::buildThreadCallGraph(SVFModule* svfModule)
+CallGraph* CallGraphBuilder::buildThreadCallGraph()
 {
-
-    buildCallGraph(svfModule);
-
+    CallGraph* callgraph = new ThreadCallGraph(*(PAG::getPAG()->getCallGraph()));
     ThreadCallGraph* cg = dyn_cast<ThreadCallGraph>(callgraph);
```
Comment (jumormt) [inline svf/lib/Util/CallGraphBuilder.cpp:69]: ThreadCallGraph* cg = new ThreadCallGraph(*(PAG::getPAG()->getCallGraph()));

File: svf/lib/Util/CallGraphBuilder.cpp [hunk 3]
```diff
@@ -56,28 +54,25 @@ CallGraph* CallGraphBuilder::buildCallGraph(SVFModule* svfModule)
...
     return callgraph;
 }
 
-CallGraph* ThreadCallGraphBuilder::buildThreadCallGraph(SVFModule* svfModule)
+CallGraph* CallGraphBuilder::buildThreadCallGraph()
 {
-
-    buildCallGraph(svfModule);
-
+    CallGraph* callgraph = new ThreadCallGraph(*(PAG::getPAG()->getCallGraph()));
```
Comment (jumormt) [inline svf/lib/Util/CallGraphBuilder.cpp:68]: delete this line

File: svf/lib/Util/CallGraphBuilder.cpp [hunk 4]
```diff
@@ -56,28 +54,25 @@ CallGraph* CallGraphBuilder::buildCallGraph(SVFModule* svfModule)
...
     ThreadCallGraph* cg = dyn_cast<ThreadCallGraph>(callgraph);
     assert(cg && "not a thread callgraph?");
 
     ThreadAPI* tdAPI = ThreadAPI::getThreadAPI();
-    for (SVFModule::const_iterator F = svfModule->begin(), E = svfModule->end(); F != E; ++F)
+    for (const auto& item: *PAG::getPAG()->getCallGraph())
```
Comment (jumormt) [inline svf/lib/Util/CallGraphBuilder.cpp:73]: use a tmp variable to store PAG::getPAG()->getCallGraph() at the beginning of the function

File: svf/lib/Util/CallGraphBuilder.cpp [hunk 5]
```diff
@@ -56,28 +54,25 @@ CallGraph* CallGraphBuilder::buildCallGraph(SVFModule* svfModule)
...
                 }
             }
         }
     }
-
     return callgraph;
 }
 
-CallGraph* ThreadCallGraphBuilder::buildThreadCallGraph(SVFModule* svfModule)
+CallGraph* CallGraphBuilder::buildThreadCallGraph()
```
Comment (jumormt) [inline svf/lib/Util/CallGraphBuilder.cpp:66]: return Threadcallgraph*

**Comment by Geoffrey1014** [comment]
--- master.txt	2024-10-11 14:19:54
+++ callgraph.txt	2024-10-11 14:19:29
@@ -31,9 +31,9 @@
 VarArrayObj         831
 VarStructObj        1664
 ----------------Time and memory stats--------------------
-LLVMIRTime          4.236
-SVFIRTime           3.074
-SymbolTableTime     0.805
+LLVMIRTime          4.189
+SVFIRTime           2.728
+SymbolTableTime     0.735
 #######################################################
 
 *********CallGraph Stats (Andersen analysis)***************
@@ -60,11 +60,11 @@
 CollapseTime        0
 CopyGepTime         0
 LoadStoreTime       0
-MemoryUsageVmrss    4.16465e+06
-MemoryUsageVmsize   4.16453e+06
+MemoryUsageVmrss    4.1647e+06
+MemoryUsageVmsize   4.16454e+06
 SCCDetectTime       0
 SCCMergeTime        0
-TotalTime           117.17
+TotalTime           118.017
 UpdateCGTime        0
 ----------------Numbers stats----------------------------
 AddrProcessed       17887
@@ -115,32 +115,32 @@
 ****Persistent Points-To Cache Statistics: Andersen's analysis bitvector****
 ################ (program : redis-server.bc)###############
 UniquePointsToSets       44837
-TotalUnions              673061
+TotalUnions              673064
 PropertyUnions           222658
 UniqueUnions             19259
 LookupUnions             395149
-PreemptiveUnions         35995
+PreemptiveUnions         35998
 TotalComplements         3364152
 PropertyComplements      2998574
 UniqueComplements        16887
 LookupComplements        331806
 PreemptiveComplements    16885
-TotalIntersections       4182813
-PropertyIntersections    4145551
-UniqueIntersections      781
-LookupIntersections      2182
-PreemptiveIntersections  34299
+TotalIntersections       4182816
+PropertyIntersections    4145395
+UniqueIntersections      851
+LookupIntersections      2268
+PreemptiveIntersections  34302
 #######################################################
 
 *********Memory SSA Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
 AverageRegSize      85.0301
-GenMUCHITime        1.757
-GenRegionTime       486.224
-InsertPHITime       0.536
-SSARenameTime       0.087
-TotalMSSATime       488.619
+GenMUCHITime        1.812
+GenRegionTime       491.501
+InsertPHITime       0.499
+SSARenameTime       0.073
+TotalMSSATime       493.903
 ----------------Numbers stats----------------------------
 BBHasMSSAPhi        10755
 CSChiNode           112337
@@ -163,13 +163,13 @@
 *********SVFG Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
-ATNodeTime          0.913
+ATNodeTime          0.9
 AvgWeight           166.005
 ConnDirEdgeTime     0
-ConnIndEdgeTime     4.459
+ConnIndEdgeTime     4.519
 OptTime             0
 TLNodeTime          0
-TotalTime           5.372
+TotalTime           5.419
 ----------------Numbers stats----------------------------
 ActualIn            164766
 ActualOut           112337
@@ -231,20 +231,20 @@
 GepTime             0
 IndirectPropaTime   0
 LoadTime            0
-MemoryUsageVmrss    3.57877e+06
-MemoryUsageVmsize   3.61642e+06
+MemoryUsageVmrss    3.57379e+06
+MemoryUsageVmsize   3.60824e+06
 PhiTime             0
 PrelabelingTime     0.126
 ProcessTime         0
 PropagationTime     0
 SCCTime             0
-SolveTime           417.791
+SolveTime           407.89
 StoreTime           0
 Strong/WeakUpdTime  0
-TotalTime           631.681
+TotalTime           624.623
 UpdateCGTime        0
 VersionPropTime     0
-meldLabelingTime    210.702
+meldLabelingTime    213.566
 ----------------Numbers stats----------------------------
 CopysNum            14138
 DummyFieldPtrs      2305
@@ -262,13 +262,13 @@
 Pointers            494739
 ProcessedAParam     0
 ProcessedAddr       107322
-ProcessedCopy       163319
+ProcessedCopy       159938
 ProcessedFRet       0
-ProcessedGep        872516
-ProcessedLoad       890456
+ProcessedGep        856747
+ProcessedLoad       850306
 ProcessedMSSANode   2318016
-ProcessedPhi        104464
-ProcessedStore      324952
+ProcessedPhi        101273
+ProcessedStore      315673
 SolveIterations     6
 StoresNum           24164
 StrongUpdates       3378
@@ -283,12 +283,12 @@
 
 ****Persistent Points-To Cache Statistics: flow-sensitive analysis bitvector****
 ################ (program : redis-server.bc)###############
-UniquePointsToSets       29750
-TotalUnions              1154903314
-PropertyUnions           997608084
-UniqueUnions             38152
-LookupUnions             157205610
-PreemptiveUnions         51468
+UniquePointsToSets       29423
+TotalUnions              1113910190
+PropertyUnions           966825643
+UniqueUnions             37254
+LookupUnions             146997055
+PreemptiveUnions         50238
 TotalComplements         0
 PropertyComplements      0
 UniqueComplements        0

File: svf/include/Util/CallGraphBuilder.h [hunk 6]
```diff
@@ -38,34 +38,30 @@ namespace SVF
...
 
 protected:
-    CallGraph* callgraph;
+    const FunctionSetType* svfFunctionSet;
     ICFG* icfg;
+
 public:
-    CallGraphBuilder(CallGraph* cg, ICFG* i): callgraph(cg),icfg(i)
+
+    CallGraphBuilder(const SVFModule::FunctionSetType* fs, ICFG* i): svfFunctionSet(fs),icfg(i)
```
Comment (jumormt) [inline svf/include/Util/CallGraphBuilder.h:54]: use FunctionSetType

File: svf/lib/Graphs/ThreadCallGraph.cpp [hunk 7]
```diff
@@ -39,9 +39,10 @@ using namespace SVFUtil;
 /*!
  * Constructor
  */
-ThreadCallGraph::ThreadCallGraph() :
-    CallGraph(ThdCallGraph), tdAPI(ThreadAPI::getThreadAPI())
+ThreadCallGraph::ThreadCallGraph(CallGraph& cg) :
+    CallGraph(cg), tdAPI(ThreadAPI::getThreadAPI())
```
Comment (jumormt) [inline svf/lib/Graphs/ThreadCallGraph.cpp:43]: kind(ThdCallGraph)

File: svf/lib/Util/CallGraphBuilder.cpp [hunk 8]
```diff
@@ -56,28 +54,25 @@ CallGraph* CallGraphBuilder::buildCallGraph(SVFModule* svfModule)
...
 }
 
-CallGraph* ThreadCallGraphBuilder::buildThreadCallGraph(SVFModule* svfModule)
+ThreadCallGraph* CallGraphBuilder::buildThreadCallGraph()
 {
-
-    buildCallGraph(svfModule);
-
-    ThreadCallGraph* cg = dyn_cast<ThreadCallGraph>(callgraph);
+    ThreadCallGraph* cg = new ThreadCallGraph(*(PAG::getPAG()->getCallGraph()));
```
Comment (jumormt) [inline svf/lib/Util/CallGraphBuilder.cpp:68]: PAG::getPAG()->getCallGraph() -> svfirCallGraph

File: svf/include/Graphs/ThreadCallGraph.h [hunk 9]
```diff
@@ -172,7 +172,7 @@ class ThreadCallGraph: public CallGraph
     typedef Map<const CallICFGNode*, ParForEdgeSet> CallInstToParForEdgesMap;
 
     /// Constructor
-    ThreadCallGraph();
+    ThreadCallGraph(CallGraph& cg);
```
Comment (yuleisui) [inline svf/include/Graphs/ThreadCallGraph.h:175]: this should be ThreadCallGraph& ?

**Pull request merged**